dspace-api/src/main/java/org/dspace/identifier/DOIIdentifierProvider.java

```diff
@@ -768,68 +765,97 @@ public String getDOIByObject(Context context, DSpaceObject dso)
      * 
      * @param context
      * @param dso The DSpaceObject the DOI should be loaded or created for.
-     * @param doi A DOI or null if a DOI should be generated. The generated DOI
+     * @param doiIdentifier A DOI or null if a DOI should be generated. The generated DOI
      * can be found in the appropriate column for the TableRow.
      * @return The database row of the object.
      * @throws SQLException In case of an error using the database.
      * @throws DOIIdentifierException If {@code doi} is not part of our prefix or
      *                             DOI is registered for another object already.
      */
-    protected DOI loadOrCreateDOI(Context context, DSpaceObject dso, String doi)
+    protected DOI loadOrCreateDOI(Context context, DSpaceObject dso, String doiIdentifier)
             throws SQLException, DOIIdentifierException
     {
-        DOI doiRow = null;
-        if (null != doi)
+        DOI doi = null;
+        if (null != doiIdentifier)
         {
             // we expect DOIs to have the DOI-Scheme except inside the doi table:
-            doi = doi.substring(DOI.SCHEME.length());
+            doiIdentifier = doiIdentifier.substring(DOI.SCHEME.length());
             
             // check if DOI is already in Database
-            doiRow = doiService.findByDoi(context, doi);
-            if (null != doiRow)
+            doi = doiService.findByDoi(context, doiIdentifier);
+            if (null != doi)
             {
-                // check if DOI already belongs to dso
-                if (ObjectUtils.equals(doiRow.getDSpaceObject(), dso))
+                if (doi.getDSpaceObject() == null)
                 {
-                    return doiRow;
-                }
-                else
-                {
-                    throw new DOIIdentifierException("Trying to create a DOI " +
-                            "that is already reserved for another object.",
-                            DOIIdentifierException.DOI_ALREADY_EXISTS);
+                    // doi was deleted, check resource type
+                    if (doi.getResourceTypeId() != null
+                            && doi.getResourceTypeId() != dso.getType())
```

Comment:
This is well handled by Java's autoboxing feature as long as doi.getResourceTypeId() is not null, which is checked in the same if clause...